internal/core/src/storage/MmapChunkManager.cpp

```diff
@@ -0,0 +1,271 @@
+    }
+    auto new_block = std::make_unique<MmapBlock>(
+        GetMmapFilePath(), size, MmapBlock::BlockType::Variable);
+    new_block->Init();
+    return std::move(new_block);
+}
+
+void
+MmapBlocksHandler::Deallocate(MmapBlockPtr&& block) {
+    if (block->GetType() == MmapBlock::BlockType::Fixed) {
+        // store the block in cache
+        block->Reset();
+        fix_size_blocks_cache_.push(std::move(block));
+        uint64_t max_cache_size =
+            uint64_t(cache_threshold * (float)max_disk_limit_);
+        if (fix_size_blocks_cache_.size() * fix_mmap_file_size_ >
+            max_cache_size) {
+            FitCache(max_cache_size);
+        }
+    } else {
+        // release the mmap
+        block->Close();
+        block = nullptr;
+    }
+}
+
+void
+MmapBlocksHandler::ClearCache() {
+    while (!fix_size_blocks_cache_.empty()) {
+        auto block = std::move(fix_size_blocks_cache_.front());
+        block->Close();
+        fix_size_blocks_cache_.pop();
+    }
+}
+
+void
+MmapBlocksHandler::FitCache(const uint64_t size) {
+    while (fix_size_blocks_cache_.size() * fix_mmap_file_size_ > size) {
+        fix_size_blocks_cache_.pop();
+    }
+}
+
+MmapChunkManager::MmapChunkManager(const std::string prefix,
+                                   const uint64_t max_limit,
+                                   const uint64_t file_size)
+    : blocks_handler_(max_limit, file_size, prefix), mmap_file_prefix_(prefix) {
+    auto cm =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    AssertInfo(cm != nullptr,
+               "Fail to get LocalChunkManager, LocalChunkManagerPtr is null");
+    if (cm->Exist(prefix)) {
+        cm->RemoveDir(prefix);
+    }
+    cm->CreateDir(prefix);
+}
+
+MmapChunkManager::~MmapChunkManager() {
+    auto cm =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    if (cm->Exist(mmap_file_prefix_)) {
+        cm->RemoveDir(mmap_file_prefix_);
+    }
+}
+
+void
+MmapChunkManager::Register(const uint64_t key) {
+    std::unique_lock<std::shared_mutex> lck(mtx_);
+    if (HasKey(key)) {
+        LOG_WARN("key has exist in growing mmap manager");
+        return;
+    }
+    blocks_table_.emplace(key, std::vector<MmapBlockPtr>());
+    return;
+}
+
+void
+MmapChunkManager::UnRegister(const uint64_t key) {
+    std::unique_lock<std::shared_mutex> lck(mtx_);
+    if (blocks_table_.find(key) != blocks_table_.end()) {
+        auto& blocks = blocks_table_[key];
+        for (auto i = 0; i < blocks.size(); i++) {
+            blocks_handler_.Deallocate(std::move(blocks[i]));
+        }
+        blocks_table_.erase(key);
+    }
+}
+
+inline bool
+MmapChunkManager::HasKey(const uint64_t key) {
+    return (blocks_table_.find(key) != blocks_table_.end());
+}
+
+void*
+MmapChunkManager::Allocate(const uint64_t key, const uint64_t size) {
+    std::unique_lock<std::shared_mutex> lck(mtx_);
```

Comment:
Let's make the gran of this lock to registered id level and take care the the thread safety of `blocks_handler_` and the cache